drivers: usb: udc: stm32: driver cleanup and EP MPS fixes

mathieuchopstm opened the pull request:
Cleans up various parts of the STM32 UDC shim driver while also fixing issues related to endpoint sizes.

mathieuchopstm committed:
drivers: usb: udc: stm32: turn EP0 max packet size into a constant

The EP0 max packet size was de facto a constant because its value was the
same regardless of which USB IP was in use. However, it was stored as part
of the instance configuration anyways which is wasteful and slower.

Create new "UDC_STM32_EP0_MAX_PACKET_SIZE" driver-level constant with which
all usage of the per-instance configuration field is replaced.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: usb: udc: stm32: clean up handling of USB buffer table

The ST USB controller (compatible 'st,stm32-usb') is fitted with private
SRAM called the Private Memory Area or PMA. This is primarily used to hold
data transfered over USB, but it also contains the so-called 'buffer table'
(a.k.a. BTABLE) which holds the base address and size of buffers in the PMA
allocated to each endpoint. The BTABLE is placed by the driver at the start
of the PMA and occupies a fixed size ('USB_BTABLE_SIZE'), which was stored
in the driver configuration field 'pma_offset'. This mechanism is unused on
non-ST USB controllers from STM32 microcontrollers, but USB_BTABLE_SIZE is
still defined (to a dummy value) and stored in the 'pma_offset' field which
becomes unused.

Remove the 'USB_BTABLE_SIZE' definition and the 'pma_offset' field from the
driver configuration, and update the ST USB controller-specific verison of
'udc_stm32_mem_init' to derive the BTABLE size from the number of endpoints
that the controller has instead.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: usb: udc: stm32: accept non-word-aligned MaxPacketSize or FIFOs

STM32 OTG USB controllers use word-addressable RAM with a 32-bit word size
so all allocations from the USB SRAM must be 32-bit aligned. The driver did
not accept unaligned values of wMaxPacketSize, and FIFO allocation code was
implicitly expecting FIFO sizes to be aligned as well (since it allocated
"bytes / 4" without rounding up).

Update driver to accept values of wMaxPacketSize that aren't word-aligned
and to allocate properly sized FIFOs sizes when an unaligned size has been
requested.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: usb: udc: stm32: allow EP max packet size up to HW capabilities

The maximal packet size (for non-control endpoints) was obtained by the
driver from HAL definitions which appear to not properly reflect hardware
capabilities.

Update driver to allow endpoints with wMaxPacketSize up to the maximal
value allowed by the USB Specification depending on operation mode, since
all STM32 USB controllers always support these values. Also move the EP
max packet size field in the 'struct udc_stm32_config' to avoid implicit
padding and add a documentation comment.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -608,7 +607,12 @@ static inline void udc_stm32_mem_init(const struct device *dev)
+	 * beginning of Private Memory Area and consumes
+	 * 8 bytes for each endpoint.
+	 */
+	priv->occupied_mem = (8 * cfg->num_endpoints);
```
Nitpicking: could drop parentheses.

mathieuchopstm replied:
My bad, copy-paste artifact from expanding `USB_BTABLE_SIZE` = `(8 * USB_NUM_BIDIR_ENDPOINTS)`.

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -121,6 +121,17 @@ LOG_MODULE_REGISTER(udc_stm32, CONFIG_UDC_DRIVER_LOG_LEVEL);
+#define UDC_STM32_NODE_EP_MPS(node_id)					\
+	((UDC_STM32_NODE_SPEED(node_id) == PCD_SPEED_HIGH) ? 1024U : 1023U)
+
+
```
Nitpicking: could you remove this extra empty line?

etienne-lms commented on drivers/usb/udc/udc_stm32.c:
```diff
@@ -1070,18 +1082,7 @@ static const struct udc_api udc_stm32_api = {
  * Kconfig system.
  */
```
Are these inline comment still useful? (non-blocking, liekly a bit outside the scope of this P-R)

mathieuchopstm replied:
IMO the comment still applies because there is some usage remaining of the Cube macros (see L1115 in new file). But eventually it should be removed yes.

etienne-lms reviewed (commented):
LGTM. Some minor comments.

mathieuchopstm committed:
drivers: usb: udc: stm32: configure OTGFS/HS RxFIFO size using Kconfig

Create a new Kconfig option allowing to tweak the RxFIFO size on OTG_FS
and OTG_HS instances, and replace the old hardcoded method with this new
mecanism.

The default value of 600 bytes yields a similar size to the the previous
hardcoded default of 160 words (= 640 bytes) when combined with the fixed
overhead computed by the driver (~56 bytes on OTG_FS with 6 endpoints).

Also fix a tiny error in a logging message (DRAM size in bytes, not bits).

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm commented:
Went for a different solution which allows tweaking RxFIFO size via Kconfig. This is acceptable for now since the driver is single-instance anyways, and more flexible than hardcoding a value in the driver.

Testing will be done at a later time to confirm no regression is observed...

mathieuchopstm commented:
> With the usual cherry-pick from 94001, I see one regression compared to [#96028 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/96028#issuecomment-3339349890): test 13 no longer passes on `nucleo_l552ze_q` (error 22, `verify_not_halted failed status -110` as usual)
> 
> This is surprising given that the board uses the ST USB IP which (should!) see 0 functional change (it isn't affected by the `RxFIFO` stuff). I'll try to see if error is present in `main`, but either way this is a non-issue to me since test 13 also fails on all boards in HS mode regardless.
> 
> Note: I did not re-test N6 since it should be almost identical to U5A5 (same IP, only PHY differs).

I also see the test 13 failure on `nucleo_l552ze_q` without these commits; something else on `main` must be causing the failures...

:arrow_right: **PR does not bring a regression** :arrow_right: OK to merge as-is

<details><summary>test details</summary>

```
$ git log --oneline
4ad0abc1578 (HEAD -> TMP__usb) [TMP] EP size patch
7eb0909f999 drivers: usb: stm32: Fix data out on control ep
0520dfe8ff9 (upstream/main, upstream/HEAD) Bluetooth: Tester: Fixed read_multiple_var
eb931d425f9 kernel/Kconfig.init: update description of SOC_RESET_HOOK
418eed0f907 arch/arm: introduce the pre-stack/RAM init hook
3d12e18b40a tests: arch: arm: irq_vector_table: don't printk from ISRs
$ ^-- cherry picked commit from #94001

$ git diff HEAD~1..HEAD
diff --git a/drivers/usb/udc/udc_stm32.c b/drivers/usb/udc/udc_stm32.c
index 7b831086c3f..f03f94350f3 100644
--- a/drivers/usb/udc/udc_stm32.c
+++ b/drivers/usb/udc/udc_stm32.c
@@ -1128,7 +1128,7 @@ static const struct udc_api udc_stm32_api = {
 
 #if defined(USB) || defined(USB_DRD_FS)
 #define EP0_MPS 64U
-#define EP_MPS 64U
+#define EP_MPS 512U
 #define USB_BTABLE_SIZE  (8 * USB_NUM_BIDIR_ENDPOINTS)
 #define USB_RAM_SIZE   DT_INST_PROP(0, ram_size)
 #else /* USB_OTG_FS */
$ ^-- mini patch to allow testusb sample to run

$ west build -b nucleo_l552ze_q samples/subsys/usb/testusb -DCONFIG_LOG_MODE_IMMEDIATE=y -DCONFIG_UDC_BUF_POOL_SIZE=16384 -DCONFIG_UDC_STM32_STACK_SIZE=16384 -DCONFIG_USBD_THREAD_STACK_SIZE=8192 -DCONFIG_ISR_STACK_SIZE=4096 -DCONFIG_HW_STACK_PROTECTION=y && west flash

$ sudo testusb -v 512 -D /dev/bus/usb/003/038 -t 13
./testusb: /dev/bus/usb/003/038 may see only control tests
full speed	/dev/bus/usb/003/038	0
running test 13
/dev/bus/usb/003/038 test 13 --> 22 (error 22)

$ sudo dmesg
[100468.649723] usbtest 3-5.3.1:1.0: TEST 13:  set/clear 1000 halts
[100478.657559] usb 3-5.3.1: verify_not_halted failed, iterations left 0, status -110 (not 0)
[100478.657568] usbtest 3-5.3.1:1.0: halts failed, iterations left 999
```

</details>

> This approach is essentially similar to what is proposed in #96206. The underlying issue is the same (need to divide memory between endpoints), and the easiest approach is to come with some sensible default and just let the user decide.

Interesting reference. There is some similarity but this seems to be a slightly different issue: issue author's instance has fixed FIFO sizes (and so their proposal describes HW parameters) whereas ST's instances have configurable FIFOs (and so our DT would be _configuration_ parameters). To be clear, I'm not against such a solution though :slightly_smiling_face:

The default for now is the historic value of 160 words(ish), though we could use Kconfig preprocessor to automatically compute e.g. `default /* 50% of zephyr_udc0 ram-size */`. (I didn't do it to minimize changes)

tmon-nordic commented:
> > This approach is essentially similar to what is proposed in #96206. The underlying issue is the same (need to divide memory between endpoints), and the easiest approach is to come with some sensible default and just let the user decide.
> 
> Interesting reference. There is some similarity but this seems to be a slightly different issue: issue author's instance has fixed FIFO sizes (and so their proposal describes HW parameters) whereas ST's instances have configurable FIFOs (and so our DT would be _configuration_ parameters). To be clear, I'm not against such a solution though 🙂

The FIFO in the referenced issue are also configurable, but not freely (the maximum length of TxFIFO is capped differently based on TxFIFO number). If the fifos were not dynamically configurable then describing the sizes in devicetree wouldn't be much useful. The whole idea behind configuring the fifo sizes at compile time is to avoid the problem of doing it dynamic at runtime (which is not trivial when there are multiple interfaces with alternate settings).

etienne-lms reviewed (approved):
Looks consistent to me.
